coderd/workspaceapps.go

```diff
@@ -235,22 +530,116 @@ func (api *API) proxyWorkspaceApplication(proxyApp proxyApplication, rw http.Res
 	proxy.ServeHTTP(rw, r)
 }
 
-// applicationCookie is a helper function to copy the auth cookie to also
-// support subdomains. Until we support creating authentication cookies that can
-// only do application authentication, we will just reuse the original token.
-// This code should be temporary and be replaced with something that creates
-// a unique session_token.
-//
-// Returns nil if the access URL isn't a hostname.
-func (api *API) applicationCookie(authCookie *http.Cookie) *http.Cookie {
-	if net.ParseIP(api.AccessURL.Hostname()) != nil {
-		return nil
-	}
-
-	appCookie := *authCookie
-	// We only support setting this cookie on the access URL subdomains. This is
-	// to ensure we don't accidentally leak the auth cookie to subdomains on
-	// another hostname.
-	appCookie.Domain = "." + api.AccessURL.Hostname()
-	return &appCookie
+type encryptedAPIKeyPayload struct {
+	APIKey    string    `json:"api_key"`
+	ExpiresAt time.Time `json:"expires_at"`
+}
+
+// encryptAPIKey encrypts an API key with it's own hashed secret. This is used
+// for smuggling (application_connect scoped) API keys securely to app
+// hostnames.
+func encryptAPIKey(data encryptedAPIKeyPayload) (string, error) {
+	if data.APIKey == "" {
+		return "", xerrors.New("API key is empty")
+	}
+	if data.ExpiresAt.IsZero() {
+		// Very short expiry as these keys are only used once as part of an
+		// automatic redirection flow.
+		data.ExpiresAt = database.Now().Add(time.Minute)
+	}
+
+	payload, err := json.Marshal(data)
+	if err != nil {
+		return "", xerrors.Errorf("marshal payload: %w", err)
+	}
+
+	// We use the hashed key secret as the encryption key. The hashed secret is
+	// stored in the API keys table. The HashedSecret is NEVER returned from the
+	// API.
+	//
+	// We chose to use the key secret as the private key for encryption instead
+	// of a shared key for a few reasons:
+	//   1. A shared key would also be stored in the database, which means that
+	//      the risk factor is similar.
+	//   2. The secret essentially rotates for each key (for free!), since each
+	//      key has a different secret. This means that if someone acquires an
+	//      old database dump they can't decrypt new API keys.
+	//   3. These tokens are scoped only for application_connect access.
+	keyID, keySecret, err := httpmw.SplitAPIToken(data.APIKey)
+	if err != nil {
+		return "", xerrors.Errorf("split API key: %w", err)
+	}
+	privateKey := sha256.Sum256([]byte(keySecret))
```

Comment:
Does this need to be hashed because of the need for a 256 bit key? If so I think a comment would be helpful.

Reply:
It's just to match the hashed secret in the database. I'll add a comment

Reply:
Is there a reason we can't just use the unhashed secret?

Reply:
We don't store the unhashed secret in the database so we wouldn't be able to decrypt the key if we did it that way. Using the hashed secret to encrypt means we can decrypt it later when the request comes in